File: frontend/tests/sync.test.ts
/**
 * Client sync correctness, exercised without a network or a browser.
 *
 * Deliberately mirrors `backend/tests/sync_test.py` case for case. The two
 * implementations have to agree about identity, conflict resolution and
 * deletion, and the cheapest way to keep them honest is to hold them to the
 * same assertions. A divergence here is a data-corruption bug that would
 * otherwise surface days later on a device that had been offline.
 *
 *     npx tsx tests/sync.test.ts
 */

import "fake-indexeddb/auto";

import {
  SYNCED_TABLES,
  type SyncRow,
  type SyncedTable,
  type Tombstone,
  clearAll,
  deleteRow,
  getRow,
  listRows,
  mergeAgentRows,
  putRows,
} from "@/lib/localdb";
import {
  type Remote,
  type SyncResult,
  moved,
  nowIso,
  summarize,
  syncOnce,
} from "@/lib/syncClient";

let passed = 0;
let failed = 0;

function check(label: string, condition: boolean, detail = ""): void {
  if (condition) {
    passed += 1;
    console.log(`  PASS  ${label}`);
  } else {
    failed += 1;
    console.log(`  FAIL  ${label}${detail ? `  -> ${detail}` : ""}`);
  }
}

/**
 * In-memory stand-in for the Supabase mirror.
 *
 * Reproduces the two behaviours the client actually depends on: upsert on
 * primary key, and a server-assigned `synced_at` that advances on every write.
 * The counter is zero-padded so lexicographic comparison orders it the way a
 * real timestamp cursor does.
 */
/** Fixed base for the fake server clock, so runs are reproducible. */
const FAKE_EPOCH = Date.UTC(2026, 0, 1);

let fakeRemoteSeq = 0;

class FakeRemote implements Remote {
  // Distinct per instance: two remotes sharing an id would share cursors
  // and bootstrap markers, which is exactly what the per-remote tests deny.
  readonly id = `fake-${(fakeRemoteSeq += 1)}`;
  rows: Record<string, Record<string, SyncRow>> = {};
  tombstones: Record<string, Tombstone & { synced_at: string }> = {};
  private clock = 0;

  constructor() {
    for (const table of SYNCED_TABLES) this.rows[table] = {};
  }

  private stamp(): string {
    // ISO rather than a bare counter: the pull cursor is rewound by a real time
    // delta before each fetch, and a counter would skip that path entirely and
    // leave the overlap window untested. The counter survives as a millisecond
    // offset purely to guarantee monotonicity.
    this.clock += 1;
    return new Date(FAKE_EPOCH + this.clock).toISOString();
  }

  /** Backdate a row's `synced_at`, as a slow transaction does.
   *
   *  Postgres stamps `now()` at transaction *start*, and transactions do not
   *  commit in start order, so a write that began earlier can become visible
   *  after one that began later. A cursor already past it would step over the
   *  row forever. */
  commitLate(table: string, uid: string, secondsEarlier: number): void {
    const row = this.rows[table][uid] as SyncRow & { synced_at: string };
    row.synced_at = new Date(Date.parse(row.synced_at) - secondsEarlier * 1000).toISOString();
  }

  private after<T extends { synced_at: string }>(
    items: T[],
    since: string | null,
  ): [T[], string | null] {
    const fresh = items
      .filter((r) => since === null || r.synced_at > since)
      .sort((a, b) => a.synced_at.localeCompare(b.synced_at));
    return [fresh, fresh.length ? fresh[fresh.length - 1].synced_at : since];
  }

  async fetchRows(table: string, since: string | null) {
    return this.after(Object.values(this.rows[table] ?? {}) as never[], since) as [
      SyncRow[],
      string | null,
    ];
  }

  async fetchTombstones(since: string | null) {
    return this.after(Object.values(this.tombstones), since) as [Tombstone[], string | null];
  }

  async pushRows(table: string, rows: SyncRow[]) {
    for (const row of rows) {
      const existing = this.rows[table][row.uid];
      // Mirrors the `keep_newer_write` trigger on the real mirror. Without it
      // the fake would be more permissive than production and the tests would
      // pass on a lie.
      if (existing && String(row.updated_at) <= String(existing.updated_at)) continue;
      this.rows[table][row.uid] = { ...row, synced_at: this.stamp() };
    }
  }

  async pushTombstones(rows: Tombstone[]) {
    for (const row of rows) {
      this.tombstones[`${row.table_name}:${row.uid}`] = { ...row, synced_at: this.stamp() };
    }
  }

  /** Write directly, bypassing the newer-wins guard, to stage states the real
   *  server would refuse — so the *client's* own guard is tested alone. */
  seed(table: string, row: SyncRow): void {
    this.rows[table][row.uid] = { ...row, synced_at: this.stamp() };
  }

  seedTombstone(table: string, uid: string, deletedAt: string): void {
    this.tombstones[`${table}:${uid}`] = {
      table_name: table,
      uid,
      deleted_at: deletedAt,
      synced_at: this.stamp(),
    };
  }
}

let uidCounter = 0;
function makeRow(extra: Record<string, unknown> = {}): SyncRow {
  uidCounter += 1;
  const stamp = nowIso();
  return {
    uid: `00000000-0000-4000-8000-${String(uidCounter).padStart(12, "0")}`,
    created_at: stamp,
    updated_at: stamp,
    ...extra,
  };
}

/** Wipe local state but keep the remote — stands in for a second device. */
async function becomeOtherDevice(): Promise<void> {
  await clearAll();
}

async function main(): Promise<number> {
  console.log("\n== local writes land and read back ==");
  const task = makeRow({ title: "Finish the sync client", priority: "HIGH", status: "PENDING" });
  const idea = makeRow({ title: "Publish SLDT", description: "after testing" });
  const event = makeRow({ event_name: "DSA revision", kind: "ROUTINE", day_of_week: 2 });
  const memory = makeRow({ key_concept: "preferred editor", content: "Neovim" });

  await putRows("tasks", [task]);
  await putRows("ideas", [idea]);
  await putRows("schedules", [event]);
  await putRows("memories", [memory]);

  check("task stored", (await listRows("tasks")).length === 1);
  check("uids are distinct", new Set([task.uid, idea.uid, event.uid, memory.uid]).size === 4);

  console.log("\n== push publishes local work ==");
  const remote = new FakeRemote();
  let result: SyncResult = await syncOnce(remote);
  check("push reported no error", result.error === null, String(result.error));
  check("task reached the remote", task.uid in remote.rows.tasks);
  check("schedule reached the remote", event.uid in remote.rows.schedules);
  check("memory reached the remote", memory.uid in remote.rows.memories);
  check("idea reached the remote", idea.uid in remote.rows.ideas);

  console.log("\n== another device pulls the same state ==");
  await becomeOtherDevice();
  check("second device starts empty", (await listRows("tasks")).length === 0);

  result = await syncOnce(remote);
  check("pull reported no error", result.error === null, String(result.error));
  const pulled = await listRows("tasks");
  check("task arrived", pulled.some((t) => t.uid === task.uid), JSON.stringify(pulled));
  check("title survived the round trip", pulled.some((t) => t.title === "Finish the sync client"));
  check("schedule arrived", (await listRows("schedules")).length === 1);
  check("memory arrived", (await listRows("memories")).length === 1);

  console.log("\n== syncing twice changes nothing ==");
  const before = (await listRows("tasks")).length;
  await syncOnce(remote);
  check("no duplicate rows", (await listRows("tasks")).length === before);

  console.log("\n== last write wins, and only when it really is later ==");
  const local = (await listRows("tasks"))[0];
  await putRows("tasks", [
    { ...local, title: "Edited locally, later", updated_at: "2098-01-01T00:00:00" },
  ]);
  remote.seed("tasks", {
    ...(remote.rows.tasks[task.uid] as SyncRow),
    title: "Stale remote edit",
    updated_at: "2020-01-01T00:00:00",
  });
  await syncOnce(remote);
  check(
    "older remote edit does not clobber a newer local one",
    (await getRow("tasks", task.uid))?.title === "Edited locally, later",
    String((await getRow("tasks", task.uid))?.title),
  );

  remote.seed("tasks", {
    ...(remote.rows.tasks[task.uid] as SyncRow),
    title: "Newer remote edit",
    updated_at: "2099-01-01T00:00:00",
  });
  await syncOnce(remote);
  check(
    "newer remote edit does land",
    (await getRow("tasks", task.uid))?.title === "Newer remote edit",
  );

  console.log("\n== deletes propagate instead of resurrecting ==");
  const doomed = makeRow({ title: "Delete me" });
  await putRows("tasks", [doomed]);
  await syncOnce(remote);
  check("doomed task published", doomed.uid in remote.rows.tasks);

  await deleteRow("tasks", doomed.uid, nowIso());
  check("delete removed it locally", (await getRow("tasks", doomed.uid)) === undefined);

  await syncOnce(remote);
  check("tombstone published", `tasks:${doomed.uid}` in remote.tombstones);

  await becomeOtherDevice();
  await syncOnce(remote);
  const surviving = new Set((await listRows("tasks")).map((t) => t.uid));
  check(
    "deleted task does not come back on the other device",
    !surviving.has(doomed.uid),
    `still present in ${[...surviving]}`,
  );
  check("the other tasks did survive", surviving.has(task.uid));

  console.log("\n== a row edited after its delete is kept ==");
  const revived = makeRow({ title: "Brought back" });
  await putRows("tasks", [revived]);
  remote.seedTombstone("tasks", revived.uid, "2020-01-01T00:00:00");
  await syncOnce(remote);
  check(
    "stale tombstone does not delete a newer row",
    (await getRow("tasks", revived.uid)) !== undefined,
  );

  console.log("\n== stale rows cannot resurrect a local deletion ==");
  const protectedRow = makeRow({ title: "Stay deleted", updated_at: "2028-01-01T00:00:00" });
  await putRows("tasks", [protectedRow]);
  await deleteRow("tasks", protectedRow.uid, "2028-01-02T00:00:00");
  remote.seed("tasks", {...protectedRow, updated_at: "2028-01-01T12:00:00"});
  await syncOnce(remote);
  check("an older pulled copy stays deleted", (await getRow("tasks", protectedRow.uid)) === undefined);

  const deliberateReturn = {...protectedRow, title: "Deliberately returned", updated_at: "2028-01-03T00:00:00"};
  await putRows("tasks", [deliberateReturn]);
  await syncOnce(remote);
  check("a later intentional edit can recreate it", (await getRow("tasks", protectedRow.uid))?.title === "Deliberately returned");

  console.log("\n== the disposable agent copy cannot overwrite newer phone data ==");
  const bridged = makeRow({ title: "Edited on phone", updated_at: "2032-01-02T00:00:00" });
  await putRows("tasks", [bridged]);
  const staleAgentCount = await mergeAgentRows("tasks", [
    { ...bridged, title: "Old agent snapshot", updated_at: "2032-01-01T00:00:00" },
  ]);
  check("older agent row is refused", staleAgentCount === 0);
  check("newer phone edit survives", (await getRow("tasks", bridged.uid))?.title === "Edited on phone");
  const freshAgentCount = await mergeAgentRows("tasks", [
    { ...bridged, title: "New agent edit", updated_at: "2032-01-03T00:00:00" },
  ]);
  check("newer agent row is accepted", freshAgentCount === 1);
  check("accepted agent edit is visible", (await getRow("tasks", bridged.uid))?.title === "New agent edit");

  console.log("\n== watermarks are per-remote ==");
  const other = new FakeRemote();
  const fresh = await syncOnce(other);
  check(
    "a different remote re-pushes in full rather than trusting a foreign cursor",
    (fresh.pushed.tasks ?? 0) > 0,
    JSON.stringify(fresh.pushed),
  );

  console.log("\n== a clock moving backwards still publishes ==");
  const backdated = makeRow({ title: "Written before the clock was corrected" });
  await putRows("tasks", [backdated]);
  await syncOnce(remote);
  // An ordinary NTP correction: the next edit carries an *earlier* stamp than
  // one already sent. A high-water-mark cursor would sort it below the mark and
  // never send it again. Whether the server then accepts it is a separate
  // question — what matters is that it was queued.
  await putRows("tasks", [
    { ...backdated, title: "Edited after the correction", updated_at: "2020-02-02T02:02:02" },
  ]);
  const afterSkew = await syncOnce(remote);
  check(
    "an edit stamped in the past is still queued and sent",
    (afterSkew.pushed.tasks ?? 0) > 0,
    "a timestamp watermark would have sorted it below the mark and skipped it",
  );

  console.log("\n== rows sharing one timestamp all publish ==");
  const twinA = makeRow({ title: "Same second, first", updated_at: "2027-03-03T03:03:03" });
  const twinB = makeRow({ title: "Same second, second", updated_at: "2027-03-03T03:03:03" });
  await putRows("tasks", [twinA, twinB]);
  await syncOnce(remote);
  check(
    "neither row is lost to a timestamp collision",
    twinA.uid in remote.rows.tasks && twinB.uid in remote.rows.tasks,
  );

  console.log("\n== a late-committing row is not stepped over ==");
  await becomeOtherDevice();
  await syncOnce(remote); // cursor advances to the newest synced_at
  remote.seed("tasks", {
    uid: "late-commit-0001",
    title: "Began early, committed late",
    created_at: "2026-05-05T05:05:05",
    updated_at: "2026-05-05T05:05:05",
  });
  // Stamped *behind* the cursor the last pull reached, exactly as a slow
  // Postgres transaction appears. Only the overlap window finds it.
  remote.commitLate("tasks", "late-commit-0001", 2);
  await syncOnce(remote);
  check(
    "a row stamped behind the cursor is still pulled",
    (await getRow("tasks", "late-commit-0001")) !== undefined,
    "the overlap window did not cover the commit-order gap",
  );

  console.log("\n== two devices editing one row converge ==");
  const contended = makeRow({ title: "Contended" });
  await putRows("tasks", [contended]);
  await syncOnce(remote);
  // The other device edits and publishes first, with the later stamp.
  remote.seed("tasks", {
    ...(remote.rows.tasks[contended.uid] as SyncRow),
    title: "Device B",
    updated_at: "2030-01-01T00:00:00",
  });
  // This device edits too, but its stamp is older.
  await putRows("tasks", [
    { ...contended, title: "Device A", updated_at: "2029-01-01T00:00:00" },
  ]);
  await syncOnce(remote);
  const settledLocal = (await getRow("tasks", contended.uid))?.title;
  const settledRemote = remote.rows.tasks[contended.uid].title;
  check(
    "both sides settle on the same version",
    settledLocal === "Device B" && settledRemote === "Device B",
    `local=${String(settledLocal)} remote=${String(settledRemote)}`,
  );

  console.log("\n== edits made offline publish on reconnect ==");
  const offline: Remote = {
    id: remote.id, // same remote, temporarily unreachable
    async fetchRows() {
      throw new Error("offline");
    },
    async pushRows() {
      throw new Error("offline");
    },
    async fetchTombstones() {
      throw new Error("offline");
    },
    async pushTombstones() {
      throw new Error("offline");
    },
  };
  const onTheTrain = makeRow({ title: "Written on a train" });
  await putRows("tasks", [onTheTrain]);
  const failedRound = await syncOnce(offline);
  check("an unreachable remote is reported, not thrown", failedRound.error !== null);
  await syncOnce(remote);
  check(
    "it publishes once the connection returns",
    onTheTrain.uid in remote.rows.tasks,
  );

  console.log("\n== sync stays out of the way when unconfigured ==");
  const idle = await syncOnce();
  check("skipped rather than threw", idle.skipped !== null, JSON.stringify(idle));
  check("reported no error", idle.error === null, String(idle.error));
  check("summary reads cleanly", summarize(idle).includes("Not synced"), summarize(idle));

  console.log("\n== reporting ==");
  check("moved() counts real movement", moved(fresh) > 0, String(moved(fresh)));

  console.log(`\n${passed} passed, ${failed} failed`);
  return failed ? 1 : 0;
}

main().then((code) => process.exit(code));

File: backend/tests/sync_test.py
"""Sync correctness, exercised end to end without touching the network.

The engine talks to a `FakeRemote` here rather than Supabase. That is not just
frugality about API credits — an in-memory remote lets a test *reorder* events
(edit after delete, stale row arriving late) which is exactly where sync goes
wrong and exactly what is impossible to stage reliably against a live server.

The live probe is a separate opt-in run:

    .venv/Scripts/python.exe tests/sync_test.py           # offline, always safe
    .venv/Scripts/python.exe tests/sync_test.py --live    # one real round trip
"""

from __future__ import annotations

import asyncio
import io
import os
import sys
import tempfile
from collections import defaultdict
from datetime import datetime, timedelta
from pathlib import Path
from typing import Any, Sequence

BACKEND = Path(__file__).resolve().parents[1]
sys.path.insert(0, str(BACKEND))
sys.stdout = io.TextIOWrapper(sys.stdout.buffer, encoding="utf-8", errors="replace")

TMP_DB = Path(tempfile.gettempdir()) / "jarvis_sync_test.db"
for suffix in ("", "-wal", "-shm"):
    target = Path(str(TMP_DB) + suffix)
    if target.exists():
        target.unlink()
os.environ["JARVIS_DB_PATH"] = str(TMP_DB)

from app.db import crud                      # noqa: E402
from app.db.database import db               # noqa: E402
from app.services import sync                # noqa: E402

#: Fixed base for the fake server clock, so runs are reproducible.
_FAKE_EPOCH = datetime(2026, 1, 1, 0, 0, 0)

PASSED = 0
FAILED = 0


def check(label: str, condition: bool, detail: str = "") -> None:
    global PASSED, FAILED
    if condition:
        PASSED += 1
        print(f"  PASS  {label}")
    else:
        FAILED += 1
        print(f"  FAIL  {label}{('  -> ' + detail) if detail else ''}")


class FakeRemote:
    """An in-memory stand-in for the Supabase mirror.

    Mimics the two behaviours the engine actually depends on: upsert-on-primary
    key, and a server-assigned `synced_at` that advances on every write. The
    counter is zero-padded so lexicographic comparison orders it correctly,
    matching how a real timestamp cursor behaves.
    """

    def __init__(self) -> None:
        self.rows: dict[str, dict[str, dict[str, Any]]] = defaultdict(dict)
        self.tombstones: dict[tuple[str, str], dict[str, Any]] = {}
        self._clock = 0
        self.closed = False

    def _stamp(self) -> str:
        """A server stamp shaped like the real one.

        ISO timestamps rather than a bare counter, because the pull cursor is
        rewound by a real time delta before each fetch. A counter would silently
        skip that code path and leave the overlap window untested. The counter
        is kept as a microsecond offset purely to guarantee monotonicity.
        """
        self._clock += 1
        return (_FAKE_EPOCH + timedelta(microseconds=self._clock)).isoformat()

    def commit_late(self, table: str, uid: str, seconds_earlier: float) -> None:
        """Backdate a row's `synced_at`, as a slow transaction does.

        Postgres stamps `now()` at transaction *start*, and transactions do not
        commit in start order. A write that began earlier can become visible
        after one that began later -- so a cursor already past it would step
        over the row and never return. This stages exactly that.
        """
        row = self.rows[table][uid]
        moment = datetime.fromisoformat(row["synced_at"]) - timedelta(seconds=seconds_earlier)
        row["synced_at"] = moment.isoformat()

    @staticmethod
    def _after(items: list[dict[str, Any]], since: str | None):
        fresh = [r for r in items if since is None or r["synced_at"] > since]
        fresh.sort(key=lambda r: r["synced_at"])
        cursor = fresh[-1]["synced_at"] if fresh else since
        return fresh, cursor

    async def fetch_rows(self, table: str, since: str | None):
        return self._after(list(self.rows[table].values()), since)

    async def fetch_tombstones(self, since: str | None):
        return self._after(list(self.tombstones.values()), since)

    async def push_rows(self, table: str, rows: Sequence[dict[str, Any]]) -> None:
        for row in rows:
            existing = self.rows[table].get(row["uid"])
            # Mirrors the keep_newer_write() trigger on the real mirror: a
            # device that has been offline must not overwrite a newer row with
            # its stale copy. Without this the fake would be more permissive
            # than production and the tests would pass on a lie.
            if existing and row["updated_at"] <= existing["updated_at"]:
                continue
            self.rows[table][row["uid"]] = {**row, "synced_at": self._stamp()}

    async def push_tombstones(self, rows: Sequence[dict[str, Any]]) -> None:
        for row in rows:
            key = (row["table_name"], row["uid"])
            self.tombstones[key] = {**row, "synced_at": self._stamp()}

    async def close(self) -> None:
        self.closed = True

    # -- helpers used by the tests, not part of the adapter protocol --------

    def seed(self, table: str, row: dict[str, Any]) -> None:
        """Write a row directly, bypassing the last-write-wins guard.

        Deliberately more permissive than `push_rows`: it stages states the real
        server would refuse, so the *client's* own guard can be tested on its
        own rather than hiding behind the server's.
        """
        self.rows[table][row["uid"]] = {**row, "synced_at": self._stamp()}

    def seed_tombstone(self, table: str, uid: str, deleted_at: str) -> None:
        self.tombstones[(table, uid)] = {
            "table_name": table,
            "uid": uid,
            "deleted_at": deleted_at,
            "synced_at": self._stamp(),
        }


async def become_other_device() -> None:
    """Wipe local rows and sync watermarks, keeping the remote intact.

    Standing in for a second machine: same code, same schema, no local state
    and nothing pulled yet.
    """
    for table in sync.SYNCED_TABLES:
        await db.execute(f"DELETE FROM {table}")
    await db.execute("DELETE FROM sync_tombstones")
    await db.execute("DELETE FROM sync_state")
    await db.execute("DELETE FROM sync_pending")


async def main(live: bool = False) -> int:
    await db.connect()

    # -- identity ---------------------------------------------------------
    print("\n== every new row gets a sync identity ==")
    task = await crud.create_task("Finish the sync engine", priority="HIGH")
    idea = await crud.create_idea("Publish SLDT", description="after testing")
    event = await crud.create_schedule_event(
        "DSA revision", kind="ROUTINE", day_of_week=2, start_time="19:30"
    )
    memory = await crud.upsert_memory("preferred editor", "Neovim", category="PREFERENCE")

    check("task has a uid", bool(task.get("uid")), repr(task.get("uid")))
    check("idea has a uid", bool(idea.get("uid")))
    check("schedule has a uid", bool(event and event.get("uid")))
    check("memory has a uid", bool(memory.get("uid")))
    check(
        "schedule carries updated_at",
        bool(event and event.get("updated_at")),
        repr(event.get("updated_at") if event else None),
    )
    check(
        "uids are distinct",
        len({task["uid"], idea["uid"], event["uid"], memory["uid"]}) == 4,
    )

    # -- push -------------------------------------------------------------
    print("\n== push publishes local work ==")
    remote = FakeRemote()
    result = await sync.sync_once(remote)
    check("push reported no error", result.error is None, str(result.error))
    check("task reached the remote", task["uid"] in remote.rows["tasks"])
    check("schedule reached the remote", event["uid"] in remote.rows["schedules"])
    check("memory reached the remote", memory["uid"] in remote.rows["memories"])
    check("idea reached the remote", idea["uid"] in remote.rows["ideas"])
    check(
        "local id did not travel",
        "id" not in remote.rows["tasks"][task["uid"]],
        "the integer id is device-local and must not sync",
    )

    # -- pull -------------------------------------------------------------
    print("\n== another device pulls the same state ==")
    await become_other_device()
    check("second device starts empty", not await crud.list_tasks())

    result = await sync.sync_once(remote)
    check("pull reported no error", result.error is None, str(result.error))
    tasks = await crud.list_tasks()
    check("task arrived", any(t["uid"] == task["uid"] for t in tasks), str(tasks))
    check(
        "title survived the round trip",
        any(t["title"] == "Finish the sync engine" for t in tasks),
    )
    check("schedule arrived", len(await crud.list_schedules()) == 1)
    check("memory arrived", len(await crud.list_memories()) == 1)

    # -- idempotence ------------------------------------------------------
    print("\n== syncing twice changes nothing ==")
    before = len(await crud.list_tasks())
    await sync.sync_once(remote)
    check("no duplicate rows", len(await crud.list_tasks()) == before)

    # -- conflicts --------------------------------------------------------
    print("\n== last write wins, and only when it really is later ==")
    local = (await crud.list_tasks())[0]
    await crud.update_task(local["id"], title="Edited locally, later")
    edited = await crud.get_task(local["id"])

    stale = dict(remote.rows["tasks"][task["uid"]])
    stale["title"] = "Stale remote edit"
    stale["updated_at"] = "2020-01-01T00:00:00"
    remote.seed("tasks", stale)

    await sync.sync_once(remote)
    now_local = await crud.get_task(local["id"])
    check(
        "older remote edit does not clobber a newer local one",
        now_local["title"] == "Edited locally, later",
        f"got {now_local['title']!r}",
    )

    fresher = dict(remote.rows["tasks"][task["uid"]])
    fresher["title"] = "Newer remote edit"
    fresher["updated_at"] = "2099-01-01T00:00:00"
    remote.seed("tasks", fresher)
    await sync.sync_once(remote)
    check(
        "newer remote edit does land",
        (await crud.get_task(local["id"]))["title"] == "Newer remote edit",
    )

    # -- deletes ----------------------------------------------------------
    print("\n== deletes propagate instead of resurrecting ==")
    doomed = await crud.create_task("Delete me")
    await sync.sync_once(remote)
    check("doomed task published", doomed["uid"] in remote.rows["tasks"])

    await crud.delete_task(doomed["id"])
    tombstones = await db.fetch_all("SELECT * FROM sync_tombstones")
    check(
        "delete recorded a tombstone",
        any(t["uid"] == doomed["uid"] for t in tombstones),
        str(tombstones),
    )

    await sync.sync_once(remote)
    check(
        "tombstone published",
        ("tasks", doomed["uid"]) in remote.tombstones,
    )

    await become_other_device()
    await sync.sync_once(remote)
    surviving = {t["uid"] for t in await crud.list_tasks()}
    check(
        "deleted task does not come back on the other device",
        doomed["uid"] not in surviving,
        f"still present in {surviving}",
    )
    check("the other tasks did survive", task["uid"] in surviving)

    # -- completing a task is a delete ------------------------------------
    print("\n== completing a task tombstones it too ==")
    finished = await crud.create_task("Ship it")
    await sync.sync_once(remote)
    await crud.update_task_status(finished["id"], "COMPLETED")
    stones = await db.fetch_all(
        "SELECT * FROM sync_tombstones WHERE uid = ?", (finished["uid"],)
    )
    check(
        "completion left a tombstone",
        bool(stones),
        "otherwise a finished task returns on the next pull",
    )

    # -- a local edit beats a stale tombstone ------------------------------
    print("\n== a row edited after its delete is kept ==")
    revived = await crud.create_task("Brought back")
    remote.seed_tombstone("tasks", revived["uid"], "2020-01-01T00:00:00")
    await sync.sync_once(remote)
    check(
        "stale tombstone does not delete a newer row",
        await crud.get_task(revived["id"]) is not None,
    )

    # -- a future timestamp must not strand later writes -------------------
    print("\n== a clock-skewed row does not freeze the push watermark ==")
    # Stands in for a second device running fast: it writes a row stamped in
    # the future, and this device pulls it. If the watermark were allowed to
    # follow that stamp, every subsequent local edit would fall below it and
    # never push again -- silently, with no error anywhere.
    skewed = await crud.create_task("Written by a fast clock")
    await db.execute(
        "UPDATE tasks SET updated_at = ? WHERE id = ?",
        ("2099-06-01T00:00:00", skewed["id"]),
    )
    await sync.sync_once(remote)

    after_skew = await crud.create_task("Written afterwards, normal clock")
    await sync.sync_once(remote)
    check(
        "a normally-stamped row still reaches the remote",
        after_skew["uid"] in remote.rows["tasks"],
        "the future-dated row dragged the watermark past the clock",
    )

    # -- backward clock skew ----------------------------------------------
    print("\n== a clock moving backwards still publishes ==")
    backdated = await crud.create_task("Written before the clock was corrected")
    await sync.sync_once(remote)
    # An ordinary NTP correction: the next edit lands with an *earlier* stamp
    # than one already sent. A high-water-mark cursor would sort it below the
    # mark and never send it again -- the failure the pending set exists to
    # remove. Whether the server then accepts it is a separate question
    # (last-write-wins says no); what matters here is that it was queued.
    await db.execute(
        "UPDATE tasks SET title = ?, updated_at = ? WHERE id = ?",
        ("Edited after the correction", "2020-02-02T02:02:02", backdated["id"]),
    )
    result = await sync.sync_once(remote)
    check(
        "an edit stamped in the past is still queued and sent",
        result.pushed.get("tasks", 0) > 0,
        "a timestamp watermark would have sorted it below the mark and skipped it",
    )

    # -- identical timestamps ----------------------------------------------
    print("\n== rows sharing one timestamp all publish ==")
    twin_a = await crud.create_task("Same second, first")
    twin_b = await crud.create_task("Same second, second")
    await db.execute(
        "UPDATE tasks SET updated_at = ? WHERE id IN (?, ?)",
        ("2027-03-03T03:03:03", twin_a["id"], twin_b["id"]),
    )
    await sync.sync_once(remote)
    check(
        "neither row is lost to a timestamp collision",
        twin_a["uid"] in remote.rows["tasks"] and twin_b["uid"] in remote.rows["tasks"],
    )

    # -- out-of-order commits ----------------------------------------------
    print("\n== a late-committing row is not stepped over ==")
    await become_other_device()
    await sync.sync_once(remote)  # cursor advances to the newest synced_at
    remote.seed(
        "tasks",
        {
            "uid": "late-commit-0001",
            "title": "Began early, committed late",
            "category": "GENERAL",
            "priority": "MEDIUM",
            "status": "PENDING",
            "due_date": None,
            "created_at": "2026-05-05T05:05:05",
            "updated_at": "2026-05-05T05:05:05",
        },
    )
    # Stamped *behind* the cursor the last pull already reached, exactly as a
    # slow Postgres transaction appears. Only the overlap window finds it.
    remote.commit_late("tasks", "late-commit-0001", seconds_earlier=2)
    await sync.sync_once(remote)
    check(
        "a row stamped behind the cursor is still pulled",
        await db.fetch_value(
            "SELECT COUNT(*) FROM tasks WHERE uid = 'late-commit-0001'", default=0
        ) == 1,
        "the overlap window did not cover the commit-order gap",
    )

    # -- concurrent writes --------------------------------------------------
    print("\n== two devices editing one row converge ==")
    contended = await crud.create_task("Contended")
    await sync.sync_once(remote)
    # The other device edits and publishes first, with the later stamp.
    remote.seed(
        "tasks",
        {**remote.rows["tasks"][contended["uid"]], "title": "Device B",
         "updated_at": "2030-01-01T00:00:00"},
    )
    # This device edits too, but its stamp is older.
    await db.execute(
        "UPDATE tasks SET title = ?, updated_at = ? WHERE id = ?",
        ("Device A", "2029-01-01T00:00:00", contended["id"]),
    )
    await sync.sync_once(remote)
    settled_local = (await crud.get_task(contended["id"]))["title"]
    settled_remote = remote.rows["tasks"][contended["uid"]]["title"]
    check(
        "both sides settle on the same version",
        settled_local == settled_remote == "Device B",
        f"local={settled_local!r} remote={settled_remote!r}",
    )

    # -- offline, then reconnect -------------------------------------------
    print("\n== edits made offline publish on reconnect ==")

    class OfflineRemote:
        """Every call fails, as an unreachable server does."""

        async def fetch_rows(self, *_: Any):
            raise RuntimeError("offline")

        async def fetch_tombstones(self, *_: Any):
            raise RuntimeError("offline")

        async def push_rows(self, *_: Any):
            raise RuntimeError("offline")

        async def push_tombstones(self, *_: Any):
            raise RuntimeError("offline")

        async def close(self) -> None:
            pass

    on_the_train = await crud.create_task("Written on a train")
    outcome = await sync.sync_once(OfflineRemote())
    check("an unreachable remote is reported, not raised", outcome.error is not None)
    check(
        "the edit is still queued after the failure",
        await db.fetch_value(
            "SELECT COUNT(*) FROM sync_pending WHERE uid = ?", (on_the_train["uid"],), default=0
        ) == 1,
    )
    await sync.sync_once(remote)
    check(
        "it publishes once the connection returns",
        on_the_train["uid"] in remote.rows["tasks"],
    )

    # -- unconfigured -----------------------------------------------------
    print("\n== sync stays out of the way when unconfigured ==")
    from app.core.config import settings

    saved = (settings.supabase_url, settings.supabase_service_key)
    settings.supabase_url, settings.supabase_service_key = "", ""
    idle = await sync.sync_once()
    settings.supabase_url, settings.supabase_service_key = saved
    check("skipped rather than raised", idle.skipped is not None, str(idle))
    check("reported no error", idle.error is None, str(idle.error))
    check("summary reads cleanly", "skipped" in idle.summary(), idle.summary())

    # -- optional live probe ----------------------------------------------
    if live:
        print("\n== live Supabase round trip ==")
        adapter = sync.build_adapter()
        if adapter is None:
            check("credentials present", False, "set SUPABASE_URL / SUPABASE_SERVICE_KEY")
        else:
            outcome = await sync.sync_once(adapter)
            check("live sync succeeded", outcome.error is None, str(outcome.error))
            print(f"        {outcome.summary()}")
            await adapter.close()

    await db.disconnect()

    print(f"\n{PASSED} passed, {FAILED} failed")
    return 1 if FAILED else 0


if __name__ == "__main__":
    raise SystemExit(asyncio.run(main(live="--live" in sys.argv)))
